File: engineering/engineering/report/oil_sample_report/oil_sample_report.js
frappe.query_reports["Oil Sample Report"] = {
	filters: [
		{
			fieldname: "location",
			label: "Location",
			fieldtype: "Link",
			options: "Location",
			description: "Leave blank for All locations",
			on_change: function () {
				// reset asset when location changes
				frappe.query_report.set_filter_value("assets", []);
				frappe.query_report.refresh();
			},
		},

		{
			fieldname: "asset_category",
			label: "Asset Category",
			fieldtype: "Link",
			options: "Asset Category",
			description: "Optional: filter asset list by category",
			on_change: function () {
				// reset assets when category changes
				frappe.query_report.set_filter_value("assets", []);
				frappe.query_report.refresh();
			},
		},


		{
			fieldname: "assets",
			label: "Assets",
			fieldtype: "MultiSelectList",
			description: "Leave empty for All assets",
			get_data: function (txt) {
				return frappe.call({
					method: "engineering.engineering.report.oil_sample_report.oil_sample_report.search_assets",
					args: {
						txt,
						location: frappe.query_report.get_filter_value("location"),
						asset_category: frappe.query_report.get_filter_value("asset_category"),
					},
				}).then((r) => r.message || []);
			},
			on_change: function () {
				frappe.query_report.refresh();
			},
		},

		{
			fieldname: "start_date",
			label: "Start Date",
			fieldtype: "Date",
			reqd: 1,
		},
		{
			fieldname: "end_date",
			label: "End Date",
			fieldtype: "Date",
			reqd: 1,
		},
		{
			fieldname: "metrics",
			label: "Value Field",
			fieldtype: "MultiSelectList",
			reqd: 1,
			default: ["fe"],
			description: "Select one or more particles (each becomes its own bar/line colour).",
			get_data: function (txt) {
				const all = [
					"tan","tbn","fe","ag","al","ca","cr","cu","mg","na","ni","pb","si","sn","p","b","ba","mo","v","zn","ti",
					"v40","v100","oxi","soot","iso4","iso6","iso14","pq","profileid"
				];
				const t = (txt || "").toLowerCase();
				return all
					.filter(x => x.toLowerCase().includes(t))
					.map(x => ({ value: x, description: x.toUpperCase() }));
			},
			on_change: function () {
				frappe.query_report.refresh();
			},
		},

		{
			fieldname: "chart_type",
			label: "Chart Type",
			fieldtype: "Select",
			default: "Bar",
			options: ["Bar", "Line"].join("\n"),
		},
		{
			fieldname: "top_n",
			label: "Top Assets Partical",
			fieldtype: "Int",
			default: 30,
			description: "Shows only the top N assets with the highest selected particle in the date range (0 = show all).",
		},
		{
			fieldname: "include_zero",
			label: "Include Zero Values",
			fieldtype: "Check",
			default: 1,
		},
	],

	onload: function (report) {
		// Pretty styling (inject once)
		if (!document.getElementById("oil-sample-report-style")) {
			const style = document.createElement("style");
			style.id = "oil-sample-report-style";
			style.innerHTML = `
				/* Title area */
				.page-title .title-text { font-weight: 700; letter-spacing: .2px; }
				/* Filters */
				.query-report .filter-section { padding: 10px 12px; border-radius: 14px; }
				/* Summary cards */
				.report-summary .summary-item {
					border-radius: 14px !important;
					box-shadow: 0 6px 18px rgba(0,0,0,.06);
				}
				/* Table */
				.datatable .dt-cell__content { font-size: 12.5px; }
				.datatable .dt-row:hover { background: rgba(0,0,0,.02); }
			`;
			document.head.appendChild(style);
		}

		// Default last 30 days
		const end = frappe.datetime.get_today();
		const start = frappe.datetime.add_days(end, -30);
		if (!report.get_filter_value("start_date")) report.set_filter_value("start_date", start);
		if (!report.get_filter_value("end_date")) report.set_filter_value("end_date", end);

	},

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Make "value" column look premium (bold + colored)
		if (column.fieldname === "value" && data) {
			const raw = Number(data.value || 0);

			// Compute max from current table rows (for relative intensity)
			const rows = (frappe.query_report && frappe.query_report.data) || [];
			let max = 0;
			for (const r of rows) max = Math.max(max, Number(r.value || 0));
			const ratio = max > 0 ? Math.min(raw / max, 1) : 0;

			// Blue → Purple scale (via HSL). Looks “world-class”.
			const hue = 220 - Math.round(40 * ratio); // 220..180
			const color = `hsl(${hue}, 80%, 40%)`;

			return `<span style="font-weight:700; color:${color};">${value}</span>`;
		}


		// Comments / Actions -> View button popup
		if ((column.fieldname === "commentstext" || column.fieldname === "actiontext") && data) {
			const title = column.fieldname === "commentstext" ? "Comments" : "Actions";
			const text = (data[column.fieldname] || "").toString().trim();

			if (!text) {
				return `<span class="text-muted">—</span>`;
			}

			// store text safely (base64) so quotes/newlines don't break HTML
			const encoded = btoa(unescape(encodeURIComponent(text)));

			return `
				<button
					class="btn btn-xs btn-default"
					data-osr-title="${title}"
					data-osr-text="${encoded}"
					onclick="window.__osr_open_text_modal(this)"
				>
					View
				</button>
			`;
		}



		// Asset emphasized
		if (column.fieldname === "asset") {
			return `<span style="font-weight:600;">${value}</span>`;
		}

		return value;
	},
};


window.__osr_open_text_modal = function (btn) {
	const title = btn.getAttribute("data-osr-title") || "Details";
	const encoded = btn.getAttribute("data-osr-text") || "";
	let text = "";
	try {
		text = decodeURIComponent(escape(atob(encoded)));
	} catch (e) {
		text = "(Could not decode text)";
	}

	const d = new frappe.ui.Dialog({
		title: title,
		size: "large",
		fields: [
			{
				fieldtype: "HTML",
				fieldname: "body",
				options: `
					<div class="osr-modal-card">
						<div class="osr-modal-badge">${frappe.utils.escape_html(title)}</div>
						<div class="osr-modal-text">${frappe.utils.escape_html(text)}</div>
					</div>
				`,
			},
		],
		primary_action_label: "Close",
		primary_action: () => d.hide(),
	});

	d.show();

	// Inject styles once + apply header gradient to this dialog
	if (!document.getElementById("osr-modal-style")) {
		const style = document.createElement("style");
		style.id = "osr-modal-style";
		style.innerHTML = `
			/* Dialog shell */
			.osr-dialog .modal-content { border-radius: 18px; overflow: hidden; box-shadow: 0 18px 45px rgba(0,0,0,.18); }
			/* Header gradient */
			.osr-dialog .modal-header {
				background: linear-gradient(90deg, rgba(37,99,235,1), rgba(124,58,237,1));
				color: #fff;
				border-bottom: none;
			}
			.osr-dialog .modal-header .modal-title { font-weight: 800; letter-spacing: .2px; }
			.osr-dialog .modal-header .close { opacity: 1; color: #fff; text-shadow: none; }
			/* Body */
			.osr-dialog .modal-body { background: #f8fafc; }
			.osr-modal-card {
				background: #fff;
				border-radius: 16px;
				padding: 14px 14px 12px 14px;
				border: 1px solid rgba(15, 23, 42, .08);
				box-shadow: 0 10px 24px rgba(0,0,0,.07);
			}
			.osr-modal-badge {
				display: inline-block;
				font-size: 11px;
				font-weight: 800;
				letter-spacing: .3px;
				padding: 6px 10px;
				border-radius: 999px;
				color: #0f172a;
				background: linear-gradient(90deg, rgba(59,130,246,.18), rgba(168,85,247,.18));
				margin-bottom: 10px;
			}
			.osr-modal-text {
				white-space: pre-wrap;
				font-size: 13px;
				line-height: 1.55;
				color: #0f172a;
				max-height: 60vh;
				overflow: auto;
				padding-right: 6px;
			}
			/* Primary button */
			.osr-dialog .modal-footer .btn-primary {
				border-radius: 12px;
				font-weight: 700;
				background: linear-gradient(90deg, rgba(37,99,235,1), rgba(124,58,237,1));
				border: none;
				box-shadow: 0 10px 20px rgba(37,99,235,.25);
			}
			.osr-dialog .modal-footer .btn-primary:hover {
				filter: brightness(1.03);
			}
		`;
		document.head.appendChild(style);
	}

	// Tag this dialog so CSS targets only our popups
	d.$wrapper.closest(".modal").addClass("osr-dialog");
};

File: engineering/engineering/report/oil_sample_report/oil_sample_report.py
import frappe
from frappe import _


ALLOWED_METRICS = {
	"tan","tbn","fe","ag","al","ca","cr","cu","mg","na","ni","pb","si","sn","p","b","ba","mo","v","zn","ti",
	"v40","v100","oxi","soot","iso4","iso6","iso14","pq","profileid"
}


def execute(filters: dict | None = None):
	filters = filters or {}

	metrics = normalize_metrics(filters.get("metrics"))
	if not metrics:
		metrics = ["fe"]

	columns = get_columns(metrics)
	rows = get_data(filters, metrics)

	chart = get_chart(rows, metrics, (filters.get("chart_type") or "Bar"))
	report_summary = get_report_summary(rows, metrics)

	return columns, rows, None, chart, report_summary


def normalize_metrics(raw) -> list[str]:
	# MultiSelectList can arrive as list or comma-separated string
	if raw is None:
		return []
	if isinstance(raw, list):
		items = raw
	else:
		items = [x.strip() for x in str(raw).split(",") if x.strip()]

	out = []
	for m in items:
		m = (m or "").strip()
		if m in ALLOWED_METRICS and m not in out:
			out.append(m)
	return out


@frappe.whitelist()
def get_locations():
	# Select filter needs a list of strings
	return [d.name for d in frappe.get_all("Location", fields=["name"], order_by="name asc")]


@frappe.whitelist()
def get_assets(location="All"):
	# If location is "All" -> return all assets
	filters = {}
	if location and location != "All":
		filters["location"] = location

	return [d.name for d in frappe.get_all("Asset", filters=filters, fields=["name"], order_by="name asc")]

def get_columns(metrics: list[str]) -> list[dict]:
	cols = [
		{"label": _("Asset"), "fieldname": "asset", "fieldtype": "Link", "options": "Asset", "width": 240},

	]

	for m in metrics:
		cols.append(
			{"label": _(m.upper()), "fieldname": f"m__{m}", "fieldtype": "Float", "width": 120}
		)

	cols += [
		{"label": _("Sample Date"), "fieldname": "sampledate", "fieldtype": "Date", "width": 120},
		{"label": _("Location"), "fieldname": "location", "fieldtype": "Data", "width": 170},
		{"label": _("Sample No"), "fieldname": "sampno", "fieldtype": "Data", "width": 140},
		{"label": _("Comments"), "fieldname": "commentstext", "fieldtype": "Small Text", "width": 260},
		{"label": _("Actions"), "fieldname": "actiontext", "fieldtype": "Small Text", "width": 260},
	]
	return cols

def normalize_assets(raw) -> list[str]:
	if not raw:
		return []
	if isinstance(raw, list):
		return [x for x in raw if x]
	return [x.strip() for x in str(raw).split(",") if x.strip()]


def get_data(filters: dict, metrics: list[str]) -> list[dict]:

	conditions = []
	params: dict = {}

	start_date = filters.get("start_date")
	end_date = filters.get("end_date")
	if not start_date or not end_date:
		end_date = frappe.utils.nowdate()
		start_date = frappe.utils.add_days(end_date, -30)

	conditions.append("sampledate between %(start_date)s and %(end_date)s")
	params["start_date"] = start_date
	params["end_date"] = end_date

	location = (filters.get("location") or "").strip()
	if location:
		conditions.append("location = %(location)s")
		params["location"] = location

	assets = normalize_assets(filters.get("assets"))
	if assets:
		conditions.append("asset in %(assets)s")
		params["assets"] = tuple(assets)



	include_zero = int(filters.get("include_zero") or 0)
	if not include_zero:
		# exclude rows where ALL selected metrics are 0
		conditions.append("(" + " or ".join([f"ifnull({m}, 0) != 0" for m in metrics]) + ")")


	where_sql = " and ".join(conditions)

	# Pick the row with the MAX of the FIRST selected metric per asset in the date range
	rank_metric = metrics[0]
	metric_select_sql = ", ".join([f"cast({m} as decimal(18,6)) as {m}" for m in metrics])

	rows = frappe.db.sql(
		f"""
		select
			t.asset,
			t.sampledate,
			t.location,
			t.sampno,
			{", ".join([f"t.{m}" for m in metrics])},
			t.commentstext,
			t.actiontext
		from (
			select
				asset,
				sampledate,
				location,
				sampno,
				{metric_select_sql},
				commentstext,
				actiontext,
				row_number() over (
					partition by asset
					order by ifnull({rank_metric}, 0) desc, sampledate desc, modified desc
				) as rn
			from `tabWearCheck Results`
			where {where_sql}
		) t
		where t.rn = 1
		order by ifnull(t.{rank_metric}, 0) desc, t.asset asc
		""",
		params,
		as_dict=True,
	)

	# map raw metric fields -> m__* so your table columns populate
	for r in rows:
		for m in metrics:
			r[f"m__{m}"] = r.get(m)



	top_n = int(filters.get("top_n") or 0)
	if top_n and top_n > 0:
		rows = rows[:top_n]

	return rows


def get_chart(rows: list[dict], metrics: list[str], chart_type: str) -> dict:
	labels = [r["asset"] for r in rows]
	ctype = "bar" if (chart_type or "").lower().startswith("b") else "line"

	datasets = []
	for m in metrics:
		datasets.append(
			{"name": m.upper(), "values": [float(r.get(m) or 0) for r in rows]}
		)

	colors = [
		"#2563EB", "#7C3AED", "#06B6D4", "#22C55E", "#F59E0B",
		"#EF4444", "#0EA5E9", "#A855F7", "#14B8A6", "#84CC16"
	]

	return {
		"data": {"labels": labels, "datasets": datasets},
		"type": ctype,
		"colors": colors[: max(1, len(datasets))],
	}



def get_report_summary(rows: list[dict], metrics: list[str]) -> list[dict]:
	first = metrics[0]
	values = [float(r.get(first) or 0) for r in rows]

	if not values:
		return [
			{"label": _("Assets"), "value": 0, "indicator": "grey"},
			{"label": _("Metrics"), "value": ", ".join([m.upper() for m in metrics]), "indicator": "blue"},
		]

	count = len(values)
	_min = min(values)
	_max = max(values)
	_avg = sum(values) / count

	return [
		{"label": _("Assets"), "value": count, "indicator": "blue"},
		{"label": _("Metrics"), "value": ", ".join([m.upper() for m in metrics]), "indicator": "blue"},
	]


